src/server/lib/okta/idx/introspect.ts

```diff
@@ -0,0 +1,143 @@
+
+const { okta } = getConfiguration();
+
+// Schema for the 'redirect-idp' object inside the introspect response remediation object
+export const redirectIdpSchema = z.object({
+	name: z.literal('redirect-idp'),
+	type: z.enum(['APPLE', 'GOOGLE']),
+	href: z.string(),
+	method: z.literal('GET'),
+});
+
+// Schema for the introspect response
+const introspectResponseSchema = z.object({
+	stateHandle: z.string(),
+	expiresAt: z.coerce.date(),
+	remediation: z.object({
+		type: z.string(),
+		value: z.array(
+			// social idp object
+			z.union([
+				redirectIdpSchema,
+				z.object({
+					// any other name thats not one of the above
+					name: z.string(),
+					type: z.string().optional(),
+					href: z.string().optional(),
+					method: z.string().optional(),
+				}),
+			]),
+		),
+	}),
+});
+export type IntrospectResponse = z.infer<typeof introspectResponseSchema>;
+
+const introspectResponseErrorSchema = z.object({
+	version: z.literal('1.0.0'),
+	messages: z.object({
+		type: z.literal('array'),
+		value: z.array(
+			z.object({
+				message: z.string(),
+				i18n: z.object({ key: z.string() }),
+			}),
+		),
+	}),
+});
+
+/**
+ * @name introspect
+ * @description Okta IDX API/Interaction Code flow - Step 2: Use the interaction handle generated from the `interact` step to start the authentication process.
+ *
+ * The introspect step lets us know what kind of authentication we can perform (remediation),
+ * and what the next steps are. It also returns the `stateHandle` which identifies the current
+ * state of the authentication process, and should be preserved and used in any subsequent requests
+ * in the flow.
+ *
+ * @param interactionHandle - The interaction handle returned from the `interact` step
+ * @param request_id - The request id
+ * @returns Promise<IntrospectResponse> - The introspect response
+ */
+export const introspect = async (
+	interactionHandle: InteractResponse['interaction_handle'],
+	request_id?: string,
+): Promise<IntrospectResponse> => {
+	try {
+		const response = await fetch(joinUrl(okta.orgUrl, '/idp/idx/introspect'), {
+			method: 'POST',
+			headers: {
+				Accept: 'application/ion+json; okta-version=1.0.0',
+				'Content-Type': 'application/ion+json; okta-version=1.0.0',
+			},
+			body: JSON.stringify({
+				interactionHandle,
+			}),
+		});
+
+		if (!response.ok) {
+			await handleError(response);
+		}
+
+		const introspectResponse = introspectResponseSchema.parse(
+			await response.json(),
+		);
+
+		trackMetric('OktaIDXIntrospect::Success');
+
+		return introspectResponse;
+	} catch (error) {
+		trackMetric('OktaIDXIntrospect::Failure');
+		logger.error('Error - Okta IDX introspect:', error, {
+			request_id,
+		});
+
+		throw error;
+	}
+};
+
+const handleError = async (response: Response) => {
+	// Check if the body is likely json using the content-type header
+	const contentType = response.headers.get('content-type');
+	if (
+		contentType &&
+		(contentType.includes('application/json') ||
```

Comment:
Could just check if `Content-Type` includes `'json'`...avoids dealing with them deciding to change to a different serialization dialect down the line (although I suspect this would be the least of our concerns at that point).

Reply:
I'd rather be explicit in the definition, as yeah if they change how they're serialising stuff, we have bigger problems 😓